essentials/src/api/executor.rs

```diff
@@ -0,0 +1,495 @@
+		request: &Request,
+		client: &dyn ApiClientT,
+	) -> color_eyre::Result<Response, RequestExecutorError> {
+		let mut retry = Retry::new(&self.retry);
+		loop {
+			match self.match_request(request.to_owned(), client).await {
+				Ok(v) => return Ok(v),
+				Err(e) => {
+					if !e.should_repeat() {
+						return Err(e)
+					}
+					if (retry.sleep().await).is_err() {
+						return Err(RequestExecutorError::Timeout)
+					}
+				},
+			}
+		}
+	}
+
+	async fn match_request(
+		&mut self,
+		request: Request,
+		client: &dyn ApiClientT,
+	) -> color_eyre::Result<Response, RequestExecutorError> {
+		use Request::*;
+		use Response::*;
+		let response = match request {
+			GetBlockTimestamp(hash) => Timestamp(client.get_block_ts(hash).await?.unwrap_or_default()),
+			GetHead(maybe_hash) => {
+				let head = match client.get_head(maybe_hash).await {
+					Ok(v) => Some(v),
+					Err(subxt::Error::Block(subxt::error::BlockError::NotFound(_))) => None,
+					Err(err) => return Err(err.into()),
+				};
+				MaybeHead(head)
+			},
+			GetBlockNumber(maybe_hash) => BlockNumber(client.get_block_number(maybe_hash).await?),
+			GetBlockHash(maybe_block_number) => MaybeBlockHash(client.legacy_get_block_hash(maybe_block_number).await?),
+			GetEvents(hash) => MaybeEvents(Some(client.get_events(hash).await?)),
+			ExtractParaInherent(maybe_hash) => ParaInherentData(client.extract_parainherent(maybe_hash).await?),
+			GetClaimQueue(hash) => {
+				let value = fetch_dynamic_storage(client, Some(hash), "ParaScheduler", "ClaimQueue").await?;
+				ClaimQueue(decode_claim_queue(&value)?)
+			},
+			GetOccupiedCores(hash) => {
+				let value = fetch_dynamic_storage(client, Some(hash), "ParaScheduler", "AvailabilityCores").await?;
+				OccupiedCores(decode_availability_cores(&value)?)
+			},
+			GetBackingGroups(hash) => {
+				let value = fetch_dynamic_storage(client, Some(hash), "ParaScheduler", "ValidatorGroups").await?;
+				BackingGroups(decode_validator_groups(&value)?)
+			},
+			GetSessionIndex(hash) => SessionIndex(client.get_session_index(hash).await?.unwrap_or_default()),
+			GetSessionAccountKeys(session_index) =>
+				SessionAccountKeys(client.get_session_account_keys(session_index).await?),
+			GetSessionNextKeys(ref account) => SessionNextKeys(client.get_session_next_keys(account).await?),
+			GetInboundHRMPChannels(hash, para_id) =>
+				HRMPChannels(client.get_inbound_hrmp_channels(hash, para_id).await?),
+			GetOutboundHRMPChannels(hash, para_id) =>
+				HRMPChannels(client.get_outbound_hrmp_channels(hash, para_id).await?),
+			GetHostConfiguration => HostConfiguration(DynamicHostConfiguration::new(
+				fetch_dynamic_storage(client, None, "Configuration", "ActiveConfig").await?,
+			)),
+			GetBestBlockSubscription => ChainSubscription(client.stream_best_block_headers().await?),
+			GetFinalizedBlockSubscription => ChainSubscription(client.stream_finalized_block_headers().await?),
+		};
+
+		Ok(response)
+	}
+}
+
+pub trait RequestExecutorNodes {
```

Comment:
I am a bit slow to understand what does the `String` represent?

Reply:
is it the rpc node url ?

Reply:
Yes, It's a bit odd way to allow pass different types of node urls as an argument.